### DISCUSSION TITLE
Loading chrome extension into QWebEngineView

### DISCUSSION THREAD

**Discussion by NiceShyGuy** [SevenTV/Extension#317] [Q&A]
Hello,

I'm working on integrating the 7tv chrome plugin with `QWebEngineView`. I have been successful in that i am able to render 7tv images in chat, I have the 7tv button which opens the emote picker, and had to inject the logo onto the button.

The problem is i keep getting this error:
`js: %c[7TV] %c[WARN]%c Could not find 7TV global stylesheet (content) color:#ef9234; color:#fac837; color:reset;`

This is because i am not loading content.js because if i load it i get the following error:
`js: Uncaught TypeError: Cannot read property 'onMessage' of undefined`

if i try to load content.js my implementation breaks and i would rather have a partially working solution then one that doesn't.

If i try to go further and eliminate this error by loading background.js i get the following error:
`js: Uncaught TypeError: Cannot read property 'onInstalled' of undefined`

I'm an amateur at all of this so i'm not sure how to resolve this issue. I'm lucky to have gotten as far as i did. If anyone can help me eliminate this error i would appreciate it. Otherwise i will just have to live with it for now.

Here is the python code that i am using currently to load the partial implementation of the chrome plugin:

```
def loadWebPlugin(me, plugin):
     if plugin == "7tv":
          plugin_path += "7tv-2.2.3/"
          me.chatView.loadFinished.connect(lambda: me.chatView.page().runJavaScript(
          f"""
              var style = document.createElement('style');
              style.innerHTML = `{open(plugin_path + "styles/Style.css", 'r').read()}`;
              document.head.appendChild(style);
          """ +
          # open(plugin_path + "background.js", "r").read() +
          # open(plugin_path + "content.js", "r").read() +
          open(plugin_path + "twitch.js", "r").read() +
          f"""
               // inject logo into .seventv-menu-button .logo
               var logo = document.createElement("img");
               logo.src = "data:image/webp;base64,{b64encode(open(plugin_path + "image/7tv.webp", "rb").read()).decode()}";
               logo.style = "height: 24px; width: 24px; margin-left: -32px; margin-top: 3px";
               // wait for button to load
               var observer = new MutationObserver(function(mutations) {{
                    mutations.forEach(function(mutation) {{
                         if (mutation.addedNodes.length) {{
                              var button = document.querySelector(".seventv-menu-button");
                              if (button) {{
                                   button.appendChild(logo);
                                   observer.disconnect();
                              }}
                         }}
                    }});
               }});
               observer.observe(document.body, {{
                    childList: true,
                    subtree: true
               }});
           """))
```

**Comment by AnatoleAM** [comment]
I wouldn't recommend doing anything with v2.2.3 as [v3.0.0](https://github.com/seventv/extensionv3) is just around the corner and it's a total rewrite. 

It's worth noting however that `content.js` and `background.js` are specific to WebExtension APIs. The content file is just a loader tailored for an Extension, so if you wish to integrate the actual site script (which is what holds the actual logic and ui for the site) you'd need to make your own loader.

**Reply by NiceShyGuy** [reply]
Thanks for clearing that up for me. I will look into how to use v3.
